File: 3. Longest Substring Without Repeating Characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/

# Given a string s, find the length of the longest substring without repeating
# characters.

###############################################################################

# longest substring -> sliding window -> use Dict for repeated char position
# Dict with key = char and val = latest_idx
# move start_idx if necessary

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s: return 0
        if len(s) == 1: return 1
        
        char_dict = dict()
        max_len = 0
        start_idx = 0 # starting index for the current substring
        for idx, char in enumerate(s):
            if char in char_dict: # found a repeated char
                char_prev_idx = char_dict[char]
                if char_prev_idx + 1 > start_idx: # move start_idx
                    # set the start idx to the right of the repeated char
                    start_idx = char_prev_idx + 1

            char_dict[char] = idx # update char idx in char_dict
                
            # update max_len
            max_len = max(max_len, idx - start_idx + 1)
            
        return max_len

File: 3. Longest Substring Without Repeating Characters.cpp
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

// Given a string s, find the length of the longest substring without repeating
// characters.

////////////////////////////////////////////////////////////////////////////////

// sliding window
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.size() == 0) return 0;
        // window[char] = count
        unordered_map<char, int> window;
        int maxLen = 0, left = 0, right = 0;       
        while (right < s.size()) {
            // update window
            char newChar = s[right];
            ++window[newChar];
            // shrink window
            while (window[newChar] > 1) {
                char oldChar = s[left];
                --window[oldChar];
                ++left;
            }
            // update maxLen
            maxLen = max(maxLen, right - left + 1);
            ++right;
        }
        return maxLen;
    }
};
